File: lib/components/images/image_url_helper.dart
/// Builds the bounded Cloudflare Images delivery contract.
class ImageUrlHelper {
  /// Standard thumbnail widths to maximize CDN cache hits.
  static const int thumbnailWidth = 300;
  static const int mediumWidth = 600;
  static const int fullWidth = 1200;

  static const Set<String> _imageHosts = {
    'img.festapp.net',
    'a.img.festapp.net',
  };
  static const Set<int> _widths = {thumbnailWidth, mediumWidth, fullWidth};

  /// Returns historical/noncanonical inputs unchanged as originals.
  static String transformImageUrl(
    String originalUrl, {
    int width = thumbnailWidth,
  }) {
    try {
      final uri = Uri.parse(originalUrl);
      final rawPath = originalUrl.split(RegExp(r'[?#]')).first;
      if (uri.scheme != 'https' ||
          uri.userInfo.isNotEmpty ||
          uri.hasPort ||
          !_imageHosts.contains(uri.host) ||
          !_widths.contains(width) ||
          !uri.path.startsWith('/images/') ||
          rawPath.contains('%') ||
          uri.pathSegments.contains('..') ||
          uri.path.startsWith('/cdn-cgi/image/') ||
          !RegExp(r'\.(jpe?g|png|webp|gif|avif|svg)$', caseSensitive: false)
              .hasMatch(uri.path)) {
        return originalUrl;
      }
      final source = Uri(scheme: 'https', host: uri.host, path: uri.path);
      final options =
          'width=$width,fit=scale-down,format=auto,quality=75,onerror=redirect';
      return 'https://${uri.host}/cdn-cgi/image/$options/$source';
    } catch (_) {
      return originalUrl;
    }
  }
}

File: web_client/src/utils/image_url_helper.js
/**
 * Builds the bounded Cloudflare Images delivery contract.
 */

const IMAGE_HOSTS = new Set(['img.festapp.net', 'a.img.festapp.net']);
const IMAGE_PATH = /\.(jpe?g|png|webp|gif|avif|svg)$/i;

/** Standard thumbnail widths to maximize CDN cache hits. */
export const THUMBNAIL_WIDTH = 300;
export const MEDIUM_WIDTH = 600;
export const FULL_WIDTH = 1200;

/**
 * Transform an img.festapp.net URL to include resize query parameters.
 * Returns original URL unchanged if not an img.festapp.net URL or on error.
 *
 * @param {string} originalUrl - The image URL to transform.
 * @param {Object} [options]
 * @param {number} [options.width=300] - Target width in pixels.
 * @returns {string} Transformed URL or original if not applicable.
 */
export function transformImageUrl(originalUrl, { width = THUMBNAIL_WIDTH } = {}) {
    try {
        const url = new URL(originalUrl);
        const rawPath = originalUrl.split(/[?#]/, 1)[0];
        if (url.protocol !== 'https:' || url.username || url.password || url.port ||
            !IMAGE_HOSTS.has(url.hostname) || ![THUMBNAIL_WIDTH, MEDIUM_WIDTH, FULL_WIDTH].includes(width) ||
            !url.pathname.startsWith('/images/') || url.pathname.startsWith('/cdn-cgi/image/') ||
            rawPath.includes('%') || url.pathname.split('/').includes('..') || !IMAGE_PATH.test(url.pathname)) {
            return originalUrl;
        }
        const source = `https://${url.hostname}${url.pathname}`;
        const options = `width=${width},fit=scale-down,format=auto,quality=75,onerror=redirect`;
        return `https://${url.hostname}/cdn-cgi/image/${options}/${source}`;
    } catch {
        return originalUrl;
    }
}
